File: src/types/signal.ts
export type SignalDirection = "long" | "short";

export type SignalStrength = "strong" | "moderate" | "weak";

export type FactorStatus = "favorable" | "neutral" | "unfavorable";

export interface SignalFactor {
  id: string;
  label: string;
  score: number; // 0-100
  status: FactorStatus;
  statusLabel: string;
  explanation: string;
}

export interface SignalScoreBreakdown {
  quant: number;
  momentum: number;
  trend: number;
  sentiment: number;
  risk: number;
}

export type SignalRating = "strong-buy" | "buy" | "neutral" | "sell" | "strong-sell";

export interface TradeSignal {
  id: string;
  symbol: string;
  name: string;
  direction: SignalDirection;
  rating: SignalRating;
  confidence: number; // 0-100
  entryLow: number;
  entryHigh: number;
  stopLoss: number;
  target: number;
  riskRewardRatio: number;
  generatedAt: string;
  scores: SignalScoreBreakdown;
  factors: SignalFactor[];
  narrative: string;
  decimals: number;
}

File: backend/app/schemas/signal.py
from datetime import datetime
from typing import Literal

from app.schemas.market import CamelModel

SignalDirection = Literal["long", "short"]
SignalRating = Literal["strong-buy", "buy", "neutral", "sell", "strong-sell"]
FactorStatus = Literal["favorable", "neutral", "unfavorable"]


class SignalFactorSchema(CamelModel):
    id: str
    label: str
    score: float
    status: FactorStatus
    status_label: str
    explanation: str


class SignalScoreBreakdownSchema(CamelModel):
    quant: float
    momentum: float
    trend: float
    sentiment: float
    risk: float


class TradeSignalSchema(CamelModel):
    id: str
    symbol: str
    name: str
    direction: SignalDirection
    rating: SignalRating
    confidence: float
    entry_low: float
    entry_high: float
    stop_loss: float
    target: float
    risk_reward_ratio: float
    generated_at: datetime
    scores: SignalScoreBreakdownSchema
    factors: list[SignalFactorSchema]
    narrative: str
    decimals: int
